internal/pkg/agent/application/upgrade/rollback_test.go

```diff
@@ -0,0 +1,507 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package upgrade
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/paths"
+	"github.com/elastic/elastic-agent/pkg/control/v2/client/mocks"
+	"github.com/elastic/elastic-agent/pkg/core/logger"
+)
+
+type hookFunc func(t *testing.T, topDir string)
+
+type agentVersion struct {
+	version string
+	hash    string
+}
+
+type agentInstall struct {
+	version          agentVersion
+	useVersionInPath bool
+}
+
+type setupAgentInstallations struct {
+	installedAgents []agentInstall
+	upgradeFrom     agentVersion
+	upgradeTo       agentVersion
+	currentAgent    agentVersion
+}
+
+var (
+	version123Snapshot = agentVersion{
+		version: "1.2.3-SNAPSHOT",
+		hash:    "abcdef",
+	}
+	version456Snapshot = agentVersion{
+		version: "4.5.6-SNAPSHOT",
+		hash:    "ghijkl",
+	}
+)
+
+func TestCleanup(t *testing.T) {
+	type args struct {
+		currentVersionedHome string
+		currentHash          string
+		removeMarker         bool
+		keepLogs             bool
+	}
+
+	tests := []struct {
```

Comment:
Nit: if you create the test cases as a `map[string]struct{...}` where the map key is the test case name, it has the minor benefit of test cases being run in arbitrary order.

Reply:
Done in 1dc1be81de77a6a689c29320a1972df9dd1683c1